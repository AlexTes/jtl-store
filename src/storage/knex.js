import envalid, { str } from 'envalid';
import knexBuilder from 'knex';
import configs from '../../knexfile';

const env = envalid.cleanEnv(process.env, {
  ENV: str({
    choices: ['production', 'staging', 'development'],
    default: 'development',
  }),
  NODE_ENV: str({
    choices: ['production', 'development', 'test'],
    default: 'development',
  }),
});

let config;
if (env.isProduction) {
  config = configs.production;
} else if (env.NODE_ENV === 'staging') {
  config = configs.staging;
} else if (env.NODE_ENV === 'test') {
  config = configs.test;
} else {
  config = configs.development;
}

export default knexBuilder(config);
