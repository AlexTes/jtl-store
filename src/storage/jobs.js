/* @flow */
import { omit } from 'lodash';
import moment from 'moment';
import knex from './knex';

export type JobProps = {
  createdAt: ?number,
  description: string,
  source: string,
  updatedAt: ?number,
}

export type RawJob = {
  createdAt: string,
  description: string,
  id: number,
  source: string,
  updatedAt: string,
}

export type Job = {
  createdAt: number,
  description: string,
  id: number,
  source: string,
  updatedAt: number,
};

export function translateFromISO(rawJob: RawJob): Job {
  return {
    ...rawJob,
    createdAt: moment.utc(rawJob.createdAt, moment.ISO_8601).format('X'),
    updatedAt: moment.utc(rawJob.updatedAt, moment.ISO_8601).format('X'),
  };
}

export function translateToISO(jobProps: JobProps) {
  return {
    ...jobProps,
    createdAt: moment.utc(jobProps.createdAt).toISOString(),
    updatedAt: moment.utc(jobProps.updatedAt).toISOString(),
  };
}

export function getOne(id: number): Promise<Job> {
  return knex('jobs')
    // Select the first safely because id has the unique constraint.
    .first()
    .where('id', id)
    .then(translateFromISO);
}

export function store(props: JobProps): Promise<number> {
  return knex('jobs')
    .insert(translateToISO(props))
    .returning('id')
    // Select the first safely because we only store one at a time
    .then(rows => rows[0]);
}
