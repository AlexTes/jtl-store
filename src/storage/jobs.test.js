// @flow
import { omit } from 'lodash';
import { getOne, store, translateToISO, translateFromISO } from './jobs';
import type { Job, RawJob } from './jobs';
import knex from './knex';

const jobProps = {
  createdAt:   1493410019,
  description: 'The developer job everyone wants.',
  source:      'Job Board',
  updatedAt:   1493410019,
};

afterAll(() => (
  knex('jobs').del().where('source', jobProps.source).then(knex.destroy)
));

test('should create a job', async () => {
  const id: number        = await store(jobProps);
  const [job]: Array<Job> = await knex('jobs').select().where('id', id);
  const jobNoID           = omit(job, 'id');

  expect(jobNoID).toMatchSnapshot();
});

test('should add defaults when needed', async () => {
  const baseJob        = omit(jobProps, 'id', 'createdAt', 'updatedAt');
  const id: number     = await store(baseJob);
  const rawJob: RawJob = await knex('jobs').select().first().where('id', id);
  const job: Job       = translateFromISO(rawJob);

  expect(job.id).toBe(id);
  expect(typeof job.createdAt).toBe('string');
  expect(typeof job.updatedAt).toBe('string');
});

test('should get one a stored job', async () => {
  const [id]: Array<number> = await knex('jobs')
    .insert(translateToISO(jobProps))
    .returning('id');
  const job     = await getOne(id);
  const jobNoID = omit(job, 'id');

  expect(job.id).toBe(id);
  expect(jobNoID).toMatchSnapshot();
});
