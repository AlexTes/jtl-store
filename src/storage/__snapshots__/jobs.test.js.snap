// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should create a job 1`] = `
Object {
  "createdAt": 1970-01-18T06:50:10.019Z,
  "description": "The developer job everyone wants.",
  "source": "Job Board",
  "updatedAt": 1970-01-18T06:50:10.019Z,
}
`;

exports[`should get one a stored job 1`] = `
Object {
  "createdAt": "1493410",
  "description": "The developer job everyone wants.",
  "source": "Job Board",
  "updatedAt": "1493410",
}
`;
